tests/UuidBinaryOrderedTimeTypeTest.php

```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    protected function setUp()
+    {
+        $this->platform = $this->getPlatformMock();
+        $this->platform->expects($this->any())
+            ->method('getBinaryTypeDeclarationSQLSnippet')
+            ->will($this->returnValue('DUMMYBINARY(16)'));
+
+        $this->type = Type::getType('uuid_binary_ordered_time');
+    }
+
+    public function testGetName()
+    {
+        $this->assertEquals('uuid_binary_ordered_time', $this->type->getName());
+    }
+
+    public function testUuidConvertsToDatabaseValue()
+    {
+        $uuid = Uuid::fromString('ff6f8cb0-c57d-11e1-9b21-0800200c9a66');
+
+        $expected = hex2bin('11e1c57dff6f8cb09b210800200c9a66');
+        $actual = $this->type->convertToDatabaseValue($uuid, $this->platform);
+
+        $this->assertEquals($expected, $actual);
+    }
+
+    public function testStringUuidConvertsToDatabaseValue()
+    {
+        $uuid = 'ff6f8cb0-c57d-11e1-9b21-0800200c9a66';
+
+        $expected = hex2bin('11e1c57dff6f8cb09b210800200c9a66');
+        $actual = $this->type->convertToDatabaseValue($uuid, $this->platform);
+
+        $this->assertEquals($expected, $actual);
+    }
+
+    public function testInvalidUuidConversionForDatabaseValue()
+    {
+        $this->setExpectedException('Doctrine\DBAL\Types\ConversionException');
+        $this->type->convertToDatabaseValue('abcdefg', $this->platform);
+    }
+
+    public function testNullConversionForDatabaseValue()
+    {
+        $this->assertNull($this->type->convertToDatabaseValue(null, $this->platform));
+    }
+
+    public function testUuidConvertsToPHPValue()
+    {
+        $uuid = $this->type->convertToPHPValue(hex2bin('11e1c57dff6f8cb09b210800200c9a66'), $this->platform);
+        $this->assertInstanceOf('Ramsey\Uuid\Uuid', $uuid);
+        $this->assertEquals('ff6f8cb0-c57d-11e1-9b21-0800200c9a66', $uuid->toString());
+    }
+
+    public function testInvalidUuidConversionForPHPValue()
+    {
+        $this->setExpectedException('Doctrine\DBAL\Types\ConversionException');
+        $this->type->convertToPHPValue('abcdefg', $this->platform);
+    }
+
+    public function testUnsupportedUuidConversionToDatabaseValue()
+    {
+        $this->setExpectedException('Doctrine\DBAL\Types\ConversionException');
+        $this->type->convertToDatabaseValue(Uuid::uuid4(), $this->platform);
+    }
+
+    /**
+     * @dataProvider provideUnsupportedDatabaseValues
+     * @param string $databaseValue
+     */
+    public function testUnsupportedUuidConversionToPHPValue($databaseValue)
+    {
+        $this->setExpectedException('Doctrine\DBAL\Types\ConversionException');
+        $this->type->convertToPHPValue(hex2bin($databaseValue), $this->platform);
+    }
+
+    public function testNullConversionForPHPValue()
+    {
+        $this->assertNull($this->type->convertToPHPValue(null, $this->platform));
+    }
+
+    public function testReturnValueIfUuidForPHPValue()
+    {
+        $uuid = Uuid::uuid4();
```

Comment:
With the current logic (passing a Uuid object to this method), this functions correctly. However, there is a potential flaw when this is binary string data coming from the database.

**There is no way to know what codec was used to encode the binary string.** This is because the codec is non-standard and has no bits set in the UUID to indicate what codec was used; RFC 4122 does not provide a means for extension.

The standard "uuid_binary" Doctrine type can encode any RFC 4122 version UUID and properly decode it, and we get the bytes back all in their proper order, and we can tell what version the UUID is, along with reading the variant, node, time, etc.

But it's possible the UUID was encoded with the "uuid_binary" Doctrine type, stored to the database, and then the code was updated to use this new "uuid_binary_ordered_time" Doctrine type. If this is the case, we have nothing in the bytes that indicates to us what codec was used, so the OrderedTimeCodec is going to do its job and rearrange the bytes back to what it thinks should be the standard order. So, any UUID that has not been encoded with the OrderedTimeCodec will be decoded improperly, with the fields out of order, and you'll be unable to grab the version, variant, time, node, etc. data from the UUID.

I don't have a solution for this. I think it's good to provide this codec, but developers will need to understand these caveats when using it. A potential "fix" (if you've found yourself in this situation where you've been generating UUIDs, storing them to the database, and now want to use this new codec) is to have a database column that indicates the specific codec used to encode the UUID, so that you can switch on the codec when decoding.